SignalServiceKit/src/Contacts/Contact.m

```diff
@@ -290,6 +290,63 @@ - (NSUInteger)hash
     return hash;
 }
 
+- (CNContact *)buildCNContactMergedWithNewContact:(CNContact *)newCNContact
+{
+    CNMutableContact *_Nullable mergedCNContact = [self.cnContact mutableCopy];
+    if (!mergedCNContact) {
+        OWSFail(@"%@ in %s mergedCNContact was unexpectedly nil", self.logTag, __PRETTY_FUNCTION__);
+        return [CNContact new];
+    }
+    
+    // Name
+    NSString *formattedFullName =  [self.class formattedFullNameWithCNContact:mergedCNContact];
+
+    // merged all or nothing - do not try to piece-meal merge.
+    if (formattedFullName.length == 0) {
+        mergedCNContact.namePrefix = newCNContact.namePrefix.ows_stripped;
+        mergedCNContact.givenName = newCNContact.givenName.ows_stripped;
+        mergedCNContact.middleName = newCNContact.middleName.ows_stripped;
+        mergedCNContact.familyName = newCNContact.familyName.ows_stripped;
+        mergedCNContact.nameSuffix = newCNContact.nameSuffix.ows_stripped;
+    }
+    
+    // Phone Numbers
+    NSSet<PhoneNumber *> *existingPhoneNumberSet = [NSSet setWithArray:self.parsedPhoneNumbers];
```

Comment:
Our system contact adapter already has a lot of tools for dealing with CNContact and their fields. So putting the merge code here was a good fit.